### ISSUE TITLE
Allow column highlight to follow caret movement into virtual space

### ISSUE THREAD

**Issue by sasumner** [vinsworldcom/nppColumnTools#2]
Hi, I recently discovered your plugin and I think it is great!  Have you considered allowing column highlight to follow caret movement into virtual space to the right of line-endings on the lines of an editing buffer?

**Comment by vinsworldcom** [comment]
Sounds cool, but I'm afraid it doesn't seem possible.  From [Scintilla docs](https://www.scintilla.org/ScintillaDoc.html):

"Virtual space is space beyond the end of each line. The caret may be moved into virtual space but no real space will be added to the document until there is some text typed or some other text insertion command is used."

**Comment by sasumner** [comment]
>but I'm afraid it doesn't seem possible

I don't see from the reference you cite anything that makes it not possible! Please reconsider. :)

**Comment by vinsworldcom** [comment]
SCI_GETCURRENTPOS → position
This returns the current position.

However, the caret moving into virtual space doesn't change the position until you actually type text.  In other words, if you have NppExec, you can test, placing your cursor at the end of some line in the middle of your doc, for example, I get:

SCI_SENDMSG SCI_GETCURRENTPOS 
================ READY ================
echo $(MSG_RESULT)
7546


Now move the cursor into virtual space way beyond the end of that same line and execute the call again:

SCI_SENDMSG SCI_GETCURRENTPOS 
================ READY ================
echo $(MSG_RESULT)
7546

7546 is the same result.  The position won't change until you actually type some characters in the virtual space - in which case the plugin responds and moves the vertical bar there.

Virtual space is just that, it's virtual, it isn't real in the sense that you're adding spaces to the end of a line until you actually type a character at the end of the virtual space you've added, which effectively inserts real ASCII spaces and then changes the current position.

As I said, I'm afraid it's not possible.

**Comment by sasumner** [comment]
Move the caret into virtual space 5 columns beyond the last real characters on a line, then:

sci_sendmsg SCI_GETSELECTIONNCARETVIRTUALSPACE 0
================ READY ================
echo $(MSG_RESULT)
5

So, one could take the "current pos" result and add in the "virtual space" result and have the column that way.  :)

**Comment by vinsworldcom** [comment]
And this SCI_GETSELECTIONNCARETVIRTUALSPACE message is available in N++ < 7.7 (with the old Scintilla version)?

**Comment by vinsworldcom** [comment]
Try this:

https://github.com/vinsworldcom/nppColumnTools/tree/develop

Let me know how it works.  Worked for me in all combos of N++ 32 / 64, versions 7.7.1 and 7.5.9.  That isn't much testing, but its both architectures, before and after the plugin update and Scintilla changes.

**Comment by sasumner** [comment]
>Try this version 1.3.3.1

Awesome, I will try it sometime today and provide feedback.

>NOTE: There are still issues (as with the current master branch version) when selecting in column mode.

Offhand I don't recall seeing any "issues" with this when selecting in column mode.  Can you enlighten me as to exactly what you are referring to so I can keep a special eye out for it during this upcoming testing?

**Comment by vinsworldcom** [comment]
Hold the Alt key down while dragging the mouse to select.  This version doesn't "break" that, it never worked.

**Comment by sasumner** [comment]
>Alt key down while dragging the mouse to select...it never worked.

Hmmm...that has always worked just fine for me in N++, with or without this plugin, although I do tend to create column selections more often with shift+alt+arrow keyboard actions.  I'll make sure to exercise the mouse way of doing it.

**Comment by vinsworldcom** [comment]
I think you misunderstood my comment.  Column selection works fine, the ruler and column marker do not function correctly - never have- when holding Alt key and selecting with mouse.

Cheers.

**Comment by vinsworldcom** [comment]
I take that back, after further experimenting, the original master version just didn't track into vspace.  Now that it does, it doesn't do it correctly with the column select (Alt-key with selection to get column selection mode) when trying to select lines of different lengths.  The ruler and column marker don't track correctly into vspace.

I think we need to look for a SCI_(column mode select) notification that needs to be taken into consideration when in that mode to move to the proper column.

**Comment by vinsworldcom** [comment]
My latest commit to develop at 9:45a EDT seems to fix the issue.  I use SCI_GETRECTANGULARSELECTIONCARETVIRTUALSPACE instead of the SCI_GETSELECTIONNCARETVIRTUALSPACE call.  The rectangular seems to work both if it's just normal select and correctly if it's a rectangular selection (column mode / hold the Alt-key while selecting).

Happy testing.  I'll be offline most of the weekend enjoying the outdoors.  Catch up on Monday.

Thanks for the issue, prodding, research and testing.

Cheers.

**Comment by sasumner** [comment]
I just started testing the earlier version and I think maybe the "N" needs to be then number of selections when a column block is being done (my original NppExec command showed N=0).  

It is acting as if it is using the first line of the blocked selection to determine the virtual space (assumes a column block being formed from top-down and left-to-right).

But maybe you are right and SCI_GETRECTANGULARSELECTIONCARETVIRTUALSPACE instead solves it.

**Comment by vinsworldcom** [comment]
Try a pull again as I just updated develop branch with that fix including
binaries version 1.3.3.2.

**Comment by sasumner** [comment]
Yes, I just tried that.  Interestingly, column mode now seems to work everywhere (in virtual space or not), but now regular non-column mode seems to have trouble when the caret is in virtual space.

**Comment by vinsworldcom** [comment]
Indeed - I just verified you're right.  Need some way to distinguish if we're in column select mode or not, then use *RETANGULARSECLECTION* or just *SELECTION*.  I'll keep looking Scintilla docs and you let me know if you find anything as well please.

**Comment by sasumner** [comment]
>way to distinguish if we're in column select mode or not

Well, SCI_SELECTIONISRECTANGLE can answer that. :)

**Comment by vinsworldcom** [comment]
I think SCI_GETSELECTIONS  does it.  just pushed to develop again, please try latest DLL has 1.3.3.3 version.

**Comment by sasumner** [comment]
> SCI_GETSELECTIONS does it

What is the "it" you are referring to?  Solving the problem? Hmm..

SCI_GETSELECTIONS will always return 1 or greater, in the case where it returns 1 it could mean several things:  a non-rectangular zero-length selection (just a normal caret), a non-rect. non-zero-length single selection, a rectangular zero-length selection (appears just like a normal caret but is NOT -- achieve by alt-left-click but no sizing drag), or a rectangular non-zero-length selection all on a single line.

In other words, I think you are going to need more logic than the addition of SCI_GETSELECTIONS in the code for this. :)

**Comment by vinsworldcom** [comment]
Did you try the 1.3.3.3 version?  It seemed to work for me.  Dropping off now.  I'll try the SCI_SELECTIONISRECTANGLE route Sunday night when I get back online.

**Comment by sasumner** [comment]
Not sure I have the latest thing.  I just downloaded it, yet it says 1.3.3.2:

![image](https://user-images.githubusercontent.com/30118311/61969392-6f688c80-afa8-11e9-88a8-85d360bec7cf.png)

Regardless, I see issues, here's a sample one, caret in col 58, edge in col 61:

![image](https://user-images.githubusercontent.com/30118311/61969516-b8204580-afa8-11e9-9e8f-ceab2bda0a66.png)

**Comment by vinsworldcom** [comment]
Yes, 1.3.3.2 won't work.  I thought I pushed 1.3.3.3 to develop before I
left.

**Comment by sasumner** [comment]
>I thought I pushed...

No worries.

I took a look at Pythonscripting this and the following logic seems to work; at least I couldn't break it.  If no problems are found with it, you could just copy the logic into C++.

    caret_virtual_space = editor.getRectangularSelectionCaretVirtualSpace() if editor.selectionIsRectangle() else editor.getSelectionNCaretVirtualSpace(editor.getMainSelection())
    editor.setEdgeColumn(editor.getColumn(editor.getCurrentPos()) + caret_virtual_space)

Of course, that means I no longer need your plugin.  :)  Just kidding, the "ruler" part is very valuable as well and there is no way I'm Pythonscripting something like that.

**Comment by vinsworldcom** [comment]
Indeed, I think this commit is 1.3.3.3:. 7775e87

To be fair, the ruler isn't mine, I found an abandoned plugin that did the ruler and added it to my vertical column plugin.  That's why the codebase looks schizophrenic.

Cheers

**Comment by vinsworldcom** [comment]
SCI_SELECTIONISRECTANGLE causes unexpected results when multiple selections (CTRL-and mouse click) are made and tracked into virtual space.  SCI_GETSELECTIONS just stops tracking in virtual space with multiple (CTRL-and mouseclick) selections.  This seems to be a more consistent stable behavior.

In both SCI_SELECTIONISRECTANGLE and SCI_GETSELECTIONS, single or rectangular track into virtual space the same.

SCI_GETSELECTIONS is the behavior in [1.3.3.3](https://github.com/vinsworldcom/nppColumnTools/releases/tag/1.3.3.3).

**Comment by vinsworldcom** [comment]
Need to use SCI_GETMAINSELECTION to pass to SCI_GETSELECTIONNCARETVIRTUALSPACE to get correct location with multiple selections.

Fixed in [1.3.3.4](https://github.com/vinsworldcom/nppColumnTools/releases/tag/1.3.3.4)

**Comment by sasumner** [comment]
I will try 1.3.3.4 later today.  I took a quick look and the logic seems to match the Pythonscript version posted above, which I haven't seen a failure in yet, so it seems promising...

**Comment by sasumner** [comment]
1.3.3.4 seems to work pretty well.  I found no issues after trying all the combinations I could think of.

**Comment by vinsworldcom** [comment]
Fixed in 1.3.3.4

**Comment by sasumner** [comment]
I noticed a new problem with this; present as well in my Pythonscript version above:

One example to reproduce is to create the following text:

    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    xxxxxxxxxxxxxxxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxx
    xxxxxxxxxxxxxxxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxx
    xxxxxxxxxxxxxxxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxx
    xxxxxxxxxxxxxxxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxx
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

Make a zero-width column caret to the left of the `-` characters.

Insert 5 spaces with 5 hits of the space bar.

Press the Del key and the vertical edge will move to a bogus point somewhere to the right of the caret.

Press Del a few more times and it just keeps getting worse.

Since this issue is closed, maybe I should make a new issue of it?

**Comment by vinsworldcom** [comment]
Seems it has to do with deleting in column mode.  That will change the SCI position and is probably messing up:

PluginDefinition.cpp:234:
void setColHi()
{
    // Get current cursor position
    HWND hCurScintilla = getCurScintilla();
    int pos = ( int )::SendMessage( hCurScintilla, SCI_GETCURRENTPOS, 0, 0 );
    int col = ( int )::SendMessage( hCurScintilla, SCI_GETCOLUMN, pos, 0 );


Wonder if there's a better way to get the column without relying on SCI_GETCURRENTPOS?  Seems the ruler indicator is not affected, just the vertical bar.

**Comment by vinsworldcom** [comment]
Think I have a fix - I'll post a test version tonight and ask you to test.  Basically, I clone the GetCaretPos() routine from the ruler part of the code into the vertical bar moving part of the code.  For my quick testing, it fixes the corner case you provided above.

Stay tuned and thanks for helping with this!

**Comment by vinsworldcom** [comment]
Please try pre-release [1.3.4.1](https://github.com/vinsworldcom/nppColumnTools/releases/tag/1.3.4.1)

**Comment by sasumner** [comment]
>Please try pre-release 1.3.4.1

Seems to work! :)

**Comment by vinsworldcom** [comment]
OK - i'll promote that pre-release to regular and close this (again).  Thanks again for the help!
